File: sdk/node/src/http.ts
/** Core HTTP client with typed error handling. */

export const DEFAULT_BASE_URL = "http://localhost:8000/api/v1";
const DEFAULT_TIMEOUT_MS = 30_000;
const LONG_TIMEOUT_MS = 300_000; // 5 min for AI/agent ops

export class APIError extends Error {
  constructor(
    public statusCode: number,
    public message: string,
    public body: Record<string, unknown> = {},
  ) {
    super(`api error ${statusCode}: ${message}`);
    this.name = "APIError";
  }
}

export class HTTPClient {
  private baseUrl: string;
  private token: string;

  constructor(baseUrl = "", token = "") {
    this.baseUrl = (baseUrl || DEFAULT_BASE_URL).replace(/\/+$/, "");
    this.token = token;
  }

  getBaseURL(): string {
    return this.baseUrl;
  }

  private async do<T>(
    method: string,
    path: string,
    opts: {
      query?: Record<string, unknown>;
      body?: unknown;
      timeout?: number;
    } = {},
  ): Promise<T> {
    const url = new URL(`${this.baseUrl}${path}`);
    if (opts.query) {
      for (const [k, v] of Object.entries(opts.query)) {
        if (v !== undefined && v !== null && v !== "" && v !== 0) {
          url.searchParams.set(k, String(v));
        }
      }
    }

    const controller = new AbortController();
    const timeout = setTimeout(
      () => controller.abort(),
      opts.timeout ?? DEFAULT_TIMEOUT_MS,
    );

    try {
      const resp = await fetch(url.toString(), {
        method,
        headers: {
          Authorization: `Bearer ${this.token}`,
          ...(opts.body ? { "Content-Type": "application/json" } : {}),
        },
        body: opts.body ? JSON.stringify(opts.body) : undefined,
        signal: controller.signal,
      });

      if (!resp.ok) {
        let body: Record<string, unknown> = {};
        try {
          body = (await resp.json()) as Record<string, unknown>;
        } catch {
          // ignore
        }
        const msg =
          (body.message as string) ?? (await resp.text()).trim();
        throw new APIError(resp.status, msg, body);
      }

      if (resp.status === 204 || resp.headers.get("content-length") === "0") {
        return undefined as T;
      }
      return (await resp.json()) as T;
    } finally {
      clearTimeout(timeout);
    }
  }

  async getJSON<T>(
    path: string,
    query?: Record<string, unknown>,
    timeout?: number,
  ): Promise<T> {
    return this.do<T>("GET", path, { query, timeout });
  }

  async postJSON<T>(
    path: string,
    body?: unknown,
    query?: Record<string, unknown>,
    timeout?: number,
  ): Promise<T> {
    return this.do<T>("POST", path, { body, query, timeout });
  }

  async putJSON<T>(
    path: string,
    body?: unknown,
    query?: Record<string, unknown>,
    timeout?: number,
  ): Promise<T> {
    return this.do<T>("PUT", path, { body, query, timeout });
  }

  async deleteJSON<T>(
    path: string,
    query?: Record<string, unknown>,
    timeout?: number,
  ): Promise<T> {
    return this.do<T>("DELETE", path, { query, timeout });
  }

  async getRaw(
    path: string,
    query?: Record<string, unknown>,
  ): Promise<Response> {
    const url = new URL(`${this.baseUrl}${path}`);
    if (query) {
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined && v !== null && v !== "" && v !== 0) {
          url.searchParams.set(k, String(v));
        }
      }
    }
    return fetch(url.toString(), {
      headers: { Authorization: `Bearer ${this.token}` },
    });
  }
}

File: sdk/python/reqmango/_http.py
"""Core HTTP client with typed error handling."""

from __future__ import annotations

from typing import Any

import httpx

DEFAULT_BASE_URL = "http://localhost:8000/api/v1"
DEFAULT_TIMEOUT = 30.0
LONG_TIMEOUT = 300.0  # 5 min for AI/agent ops


class APIError(Exception):
    """Typed error carrying the backend's {"message": ...} body."""

    def __init__(self, status_code: int, message: str, body: dict[str, Any] | None = None):
        self.status_code = status_code
        self.message = message
        self.body = body or {}
        super().__init__(f"api error {status_code}: {message}")


class HTTPClient:
    """Low-level HTTP client shared by all resource mixins."""

    def __init__(self, base_url: str = "", token: str = "", timeout: float = DEFAULT_TIMEOUT):
        self._base_url = (base_url or DEFAULT_BASE_URL).rstrip("/")
        self._token = token
        self._client = httpx.Client(
            timeout=timeout,
            headers={"Authorization": f"Bearer {token}"},
        )

    @property
    def base_url(self) -> str:
        return self._base_url

    def _do(
        self,
        method: str,
        path: str,
        *,
        query: dict[str, Any] | None = None,
        body: Any = None,
        out_type: type | None = None,
        timeout: float | None = None,
    ) -> Any:
        url = f"{self._base_url}{path}"
        kwargs: dict[str, Any] = {}
        if query:
            kwargs["params"] = {k: v for k, v in query.items() if v is not None and v != "" and v != 0}
        if body is not None:
            kwargs["json"] = body
        if timeout:
            kwargs["timeout"] = timeout

        resp = self._client.request(method, url, **kwargs)

        if resp.status_code < 200 or resp.status_code >= 300:
            body_data: dict[str, Any] = {}
            try:
                body_data = resp.json()
            except Exception:
                pass
            msg = body_data.get("message", resp.text.strip())
            raise APIError(resp.status_code, msg, body_data)

        if resp.status_code == 204 or not resp.content:
            return None
        data = resp.json()
        return data

    def get_json(self, path: str, query: dict[str, Any] | None = None, *, timeout: float | None = None) -> Any:
        return self._do("GET", path, query=query, timeout=timeout)

    def post_json(self, path: str, body: Any = None, query: dict[str, Any] | None = None, *, timeout: float | None = None) -> Any:
        return self._do("POST", path, query=query, body=body, timeout=timeout)

    def put_json(self, path: str, body: Any = None, query: dict[str, Any] | None = None, *, timeout: float | None = None) -> Any:
        return self._do("PUT", path, query=query, body=body, timeout=timeout)

    def delete_json(self, path: str, query: dict[str, Any] | None = None, *, timeout: float | None = None) -> Any:
        return self._do("DELETE", path, query=query, timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> HTTPClient:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()
